### ISSUE TITLE
pscnv.ko fails to load (valinna linux 2.6.38.8 kernel, GTX580 GPU)

### ISSUE THREAD

**Issue by wkbjerry** [shinpei0208/gdev#7]
The pscnv in gdev compiles well, but fails to load on my machine with valinna 2.6.38.8 kernel. The dmesg output shows:

[drm] Initialized drm 1.1.0 20060810
pscnv 0000:01:00.0: PCI INT A -> GSI 16 (level, low) -> IRQ 16
pscnv 0000:01:00.0: setting latency timer to 64
[drm] pscnv 0000:01:00.0: Detected an NVc0 generation card (0x0c8000a1)
[drm] pscnv 0000:01:00.0: Initializing card...
[drm] pscnv 0000:01:00.0: Attempting to load BIOS image from PRAMIN
[drm] pscnv 0000:01:00.0: ... appears to be valid
[drm] pscnv 0000:01:00.0: BIT BIOS found
[drm] pscnv 0000:01:00.0: Bios version 70.10.48.00
[drm] pscnv 0000:01:00.0: TMDS table version 2.0
[drm] pscnv 0000:01:00.0: Found Display Configuration Block version 4.0
[drm] pscnv 0000:01:00.0: Raw DCB entry 0: 02000300 00000000
[drm] pscnv 0000:01:00.0: Raw DCB entry 1: 01000302 00020030
[drm] pscnv 0000:01:00.0: Raw DCB entry 2: 04011380 00000000
[drm] pscnv 0000:01:00.0: Raw DCB entry 3: 08011382 00020030
[drm] pscnv 0000:01:00.0: Raw DCB entry 4: 02022362 00020010
[drm] pscnv 0000:01:00.0: DCB connector table: VHER 0x40 5 16 4
[drm] pscnv 0000:01:00.0:   0: 0x00001030: type 0x30 idx 0 tag 0x07
[drm] pscnv 0000:01:00.0:   1: 0x00010130: type 0x30 idx 1 tag 0x51
[drm] pscnv 0000:01:00.0:   2: 0x00002261: type 0x61 idx 2 tag 0x08
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table 0 at offset 0x6BD4
[drm] pscnv 0000:01:00.0: unknown i2c port 51
[drm] pscnv 0000:01:00.0: 0x6B49: Failed parsing init table opcode: INIT_I2C_IF -19
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table 1 at offset 0x7296
[drm] pscnv 0000:01:00.0: unknown i2c port 51
[drm] pscnv 0000:01:00.0: 0x8553: Failed parsing init table opcode: INIT_I2C_LONG_IF -19
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table 2 at offset 0x878C
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table 3 at offset 0x8796
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table 4 at offset 0x8A0A
[drm] pscnv 0000:01:00.0: Parsing VBIOS init table at offset 0x8A6F
[drm] pscnv 0000:01:00.0: 0x8A6F: Condition still not met after 20ms, skipping following opcodes
[drm] pscnv 0000:01:00.0: VRAM: size 0x60000000, 6 controllers
[drm] pscnv 0000:01:00.0: unknown i2c port 51
BUG: unable to handle kernel NULL pointer dereference at 0000000000000010
IP: [<ffffffffa00b92b6>] i2c_transfer+0x16/0x110 [i2c_core]
PGD 22c58e067 PUD 22de76067 PMD 0 
Oops: 0000 [#1] SMP 
last sysfs file: /sys/module/i2c_core/initstate
CPU 0 
Modules linked in: pscnv(+) drm_kms_helper drm i2c_algo_bit video autofs4 ipt_REJECT nf_conntrack_ipv4 nf_defrag_ipv4 iptable_filter ip_tables nf_conntrack_tftp nf_conntrack_ftp ip6t_REJECT ip6t_ipv6header nf_conntrack_ipv6 nf_defrag_ipv6 xt_state nf_conntrack ip6table_filter ip6_tables ipv6 dm_mirror dm_region_hash dm_log dm_mod uinput sg dcdbas microcode pcspkr serio_raw i2c_i801 i2c_core iTCO_wdt iTCO_vendor_support broadcom tg3 snd_hda_codec_hdmi snd_hda_codec_realtek snd_hda_intel snd_hda_codec snd_hwdep snd_seq snd_seq_device snd_pcm snd_timer snd soundcore snd_page_alloc ext4 mbcache jbd2 sd_mod crc_t10dif ahci libahci [last unloaded: speedstep_lib]

Pid: 2010, comm: modprobe Not tainted 2.6.38.8 #1 Dell Inc.  Precision T1500/0P67HD
RIP: 0010:[<ffffffffa00b92b6>]  [<ffffffffa00b92b6>] i2c_transfer+0x16/0x110 [i2c_core]
RSP: 0018:ffff880229da3828  EFLAGS: 00010296
RAX: ffff880229da38a7 RBX: 0000000000000000 RCX: 0000000000003c75
RDX: 0000000000000002 RSI: ffff880229da3878 RDI: 0000000000000000
RBP: ffff880229da3868 R08: 0000000000000000 R09: ffffffff816419e0
R10: 0000000000000001 R11: 0000000000000000 R12: ffffffffa038d4e0
R13: ffff880229da3908 R14: 0000000000000000 R15: ffff88022c70f000
FS:  00007f919cd84700(0000) GS:ffff8800bf400000(0000) knlGS:0000000000000000
CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
CR2: 0000000000000010 CR3: 000000022c0f6000 CR4: 00000000000006f0
DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
DR3: 0000000000000000 DR6: 00000000ffff0ff0 DR7: 0000000000000400
Process modprobe (pid: 2010, threadinfo ffff880229da2000, task ffff88022ca8ab30)
Stack:
 ffffffffa03ea5b2 0000000000000000 ffff88022c70f000 0000000000000000
 ffffffffa038d4e0 ffff880229da3908 0000000000000000 ffff88022c70f000
 ffff880229da38a8 ffffffffa038487b ffff00010000002d ffff880229da38a7
Call Trace:
 [<ffffffffa038d4e0>] ? probe_monitoring_device+0x0/0xc0 [pscnv]
 [<ffffffffa038487b>] nouveau_probe_i2c_addr+0x4b/0x60 [pscnv]
 [<ffffffffa038505b>] nouveau_i2c_identify+0x5b/0x170 [pscnv]
 [<ffffffffa038d03d>] nouveau_temp_probe_i2c+0xdd/0xe0 [pscnv]
 [<ffffffffa038d06d>] nouveau_temp_init+0x2d/0x4a0 [pscnv]
 [<ffffffffa038b1e1>] nouveau_pm_init+0x51/0x4f0 [pscnv]
 [<ffffffff8105bcda>] ? __cond_resched+0x2a/0x40
 [<ffffffff814b1420>] ? _cond_resched+0x30/0x40
 [<ffffffff814b1f7e>] ? mutex_lock+0x1e/0x50
 [<ffffffff81065202>] ? put_online_cpus+0x52/0x70
 [<ffffffff810257a5>] ? mtrr_add_page+0x1b5/0x470
 [<ffffffff81025ab6>] ? mtrr_add+0x56/0x70
 [<ffffffffa036a716>] nouveau_card_init+0x206/0x8c0 [pscnv]
 [<ffffffffa036b420>] nouveau_load+0x570/0x8a0 [pscnv]
 [<ffffffffa0336770>] drm_get_pci_dev+0x180/0x2a0 [drm]
 [<ffffffffa03e3f15>] nouveau_pci_probe+0x15/0x1c [pscnv]
 [<ffffffff81249a2f>] local_pci_probe+0x5f/0xd0
 [<ffffffff8124ad21>] pci_device_probe+0x101/0x120
 [<ffffffff81309bda>] ? driver_sysfs_add+0x7a/0xb0
 [<ffffffff81309d73>] driver_probe_device+0xa3/0x2b0
 [<ffffffff8130a02b>] __driver_attach+0xab/0xb0
 [<ffffffff81309f80>] ? __driver_attach+0x0/0xb0
 [<ffffffff813090d4>] bus_for_each_dev+0x64/0x90
 [<ffffffff81309b0e>] driver_attach+0x1e/0x20
 [<ffffffff813094f8>] bus_add_driver+0x1e8/0x290
 [<ffffffffa0409000>] ? nouveau_init+0x0/0xd4 [pscnv]
 [<ffffffff8130a366>] driver_register+0x76/0x140
 [<ffffffffa0409000>] ? nouveau_init+0x0/0xd4 [pscnv]
 [<ffffffff8124af76>] __pci_register_driver+0x56/0xd0
 [<ffffffffa0336974>] drm_pci_init+0xe4/0xf0 [drm]
 [<ffffffffa0409000>] ? nouveau_init+0x0/0xd4 [pscnv]
 [<ffffffffa032e438>] drm_init+0x58/0x70 [drm]
 [<ffffffffa04090cf>] nouveau_init+0xcf/0xd4 [pscnv]
 [<ffffffff81002053>] do_one_initcall+0x43/0x190
 [<ffffffff810a008b>] sys_init_module+0xbb/0x200
 [<ffffffff8100bf82>] system_call_fastpath+0x16/0x1b
Code: 00 00 00 00 00 48 83 e8 58 74 e6 48 89 c7 e8 c2 ff ff ff c9 c3 55 48 89 e5 41 57 41 56 41 55 41 54 53 48 83 ec 18 0f 1f 44 00 00 <48> 8b 47 10 48 89 fb 49 89 f5 41 89 d6 48 83 38 00 0f 84 a8 00 
RIP  [<ffffffffa00b92b6>] i2c_transfer+0x16/0x110 [i2c_core]
 RSP <ffff880229da3828>
CR2: 0000000000000010
---[ end trace c26142068be8f301 ]---

I checked modules.dep. The modules pscnv.ko depend on are:
extra/pscnv.ko: kernel/drivers/gpu/drm/drm_kms_helper.ko kernel/drivers/gpu/drm/drm.ko kernel/drivers/i2c/algos/i2c-algo-bit.ko kernel/drivers/i2c/i2c-core.ko kernel/drivers/acpi/video.ko
All from the valinna linux-2.6.38.8 installation. I guess there may be something wrong within some of the modules. 

Maybe I should try to compile a history 2.6.38.8 kernel fetched from the nouveau repository (git://anongit.freedesktop.org/nouveau/linux-2.6), rather than from the valinna linux kernel website (www.kernel.org)?

**Comment by shinpei0208** [comment]
Kaibo,

What is your GPU? pscnv doesn't work with ones except for nvc0.

**Comment by wkbjerry** [comment]
The GPU is GTX580, which is nvc8. This may be the reason. I will try again
after my GTX480 arrives very soon.

**Comment by shinpei0208** [comment]
Or try using Nouveau. Nouveau is available for almost all Fermi GPUs.

**Comment by wkbjerry** [comment]
I tried the Nouveau bundled with gdev on 2.6.38.8, but it failed to
compile. Do you have a nouveau patch for gdev on 2.6.38.8? Otherwise, I'll
have to wait for the GTX 480 and use pscnv.

**Comment by shinpei0208** [comment]
Doesn't gdev/mod/patches/gdev-nouveau-2.6.38.8.patche work?

**Comment by wkbjerry** [comment]
The patch does not work with the nouveau source code bundled with vanilla linux 2.6.38.8. Here's the message when applying the patch:

patching file drivers/gpu/drm/nouveau/Makefile
patching file drivers/gpu/drm/nouveau/gdev_interface.c
patching file drivers/gpu/drm/nouveau/gdev_interface.h
patching file drivers/gpu/drm/nouveau/nouveau_channel.c
Hunk #1 FAILED at 188.
1 out of 1 hunk FAILED -- saving rejects to file drivers/gpu/drm/nouveau/nouveau_channel.c.rej
patching file drivers/gpu/drm/nouveau/nouveau_drv.c
Hunk #1 succeeded at 432 with fuzz 1 (offset -28 lines).
patching file drivers/gpu/drm/nouveau/nouveau_state.c
Hunk #1 succeeded at 909 (offset -81 lines).
Hunk #2 succeeded at 1051 (offset -113 lines).
patching file drivers/gpu/drm/nouveau/nv50_vm.c
Hunk #1 FAILED at 85.
1 out of 1 hunk FAILED -- saving rejects to file drivers/gpu/drm/nouveau/nv50_vm.c.rej
patching file drivers/gpu/drm/nouveau/nvc0_graph.c
Hunk #1 succeeded at 716 with fuzz 1 (offset 54 lines).
Hunk #2 succeeded at 731 with fuzz 2 (offset 54 lines).

The problem at this step seems to be about nv50_vm.c.
    phys  = vm_addr(vma, phys, mem->memtype, 0);
    pte <<= 3;
    cnt <<= 3;
is expected, but what's in drivers/gpu/drm/nouveau/nv50_vm.c is:
    phys  = nv50_vm_addr(vma, pgt, phys, mem->memtype, 0);
    pte <<= 3;
    cnt <<= 3;

Then, here's the error when compiling the kernel:

  CC [M]  drivers/gpu/drm/nouveau/gdev_interface.o
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_chan_alloc’:
drivers/gpu/drm/nouveau/gdev_interface.c:58: error: ‘struct nouveau_channel’ has no member named ‘pushbuf_vma’
drivers/gpu/drm/nouveau/gdev_interface.c:79: error: ‘struct drm_nouveau_private’ has no member named ‘eng’
drivers/gpu/drm/nouveau/gdev_interface.c:79: error: ‘NVOBJ_ENGINE_COPY0’ undeclared (first use in this function)
drivers/gpu/drm/nouveau/gdev_interface.c:79: error: (Each undeclared identifier is reported only once
drivers/gpu/drm/nouveau/gdev_interface.c:79: error: for each function it appears in.)
drivers/gpu/drm/nouveau/gdev_interface.c:82: error: ‘struct drm_nouveau_private’ has no member named ‘eng’
drivers/gpu/drm/nouveau/gdev_interface.c:82: error: ‘NVOBJ_ENGINE_COPY1’ undeclared (first use in this function)
drivers/gpu/drm/nouveau/gdev_interface.c:110: error: ‘struct drm_nouveau_private’ has no member named ‘eng’
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_chan_free’:
drivers/gpu/drm/nouveau/gdev_interface.c:121: error: ‘struct drm_nouveau_private’ has no member named ‘eng’
drivers/gpu/drm/nouveau/gdev_interface.c:121: error: ‘NVOBJ_ENGINE_COPY1’ undeclared (first use in this function)
drivers/gpu/drm/nouveau/gdev_interface.c:122: error: ‘struct drm_nouveau_private’ has no member named ‘eng’
drivers/gpu/drm/nouveau/gdev_interface.c:122: error: ‘NVOBJ_ENGINE_COPY0’ undeclared (first use in this function)
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_bo_alloc’:
drivers/gpu/drm/nouveau/gdev_interface.c:145: warning: passing argument 2 of ‘nouveau_bo_new’ makes pointer from integer without a cast
drivers/gpu/drm/nouveau/nouveau_drv.h:1295: note: expected ‘struct nouveau_channel _’ but argument is of type ‘uint64_t’
drivers/gpu/drm/nouveau/gdev_interface.c:145: warning: passing argument 7 of ‘nouveau_bo_new’ makes integer from pointer without a cast
drivers/gpu/drm/nouveau/nouveau_drv.h:1295: note: expected ‘uint32_t’ but argument is of type ‘struct nouveau_bo *_’
drivers/gpu/drm/nouveau/gdev_interface.c:145: error: too few arguments to function ‘nouveau_bo_new’
drivers/gpu/drm/nouveau/gdev_interface.c:166: error: implicit declaration of function ‘nouveau_bo_vma_add’
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_bo_free’:
drivers/gpu/drm/nouveau/gdev_interface.c:212: error: implicit declaration of function ‘nouveau_bo_vma_find’
drivers/gpu/drm/nouveau/gdev_interface.c:212: warning: assignment makes pointer from integer without a cast
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_bo_unbind’:
drivers/gpu/drm/nouveau/gdev_interface.c:273: warning: assignment makes pointer from integer without a cast
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_getparam’:
drivers/gpu/drm/nouveau/gdev_interface.c:390: error: implicit declaration of function ‘nv_engine’
drivers/gpu/drm/nouveau/gdev_interface.c:390: warning: initialization makes pointer from integer without a cast
drivers/gpu/drm/nouveau/gdev_interface.c: In function ‘gdev_drv_getaddr’:
drivers/gpu/drm/nouveau/gdev_interface.c:429: error: dereferencing pointer to incomplete type
make[4]: **\* [drivers/gpu/drm/nouveau/gdev_interface.o] Error 1
make[3]: **\* [drivers/gpu/drm/nouveau] Error 2
make[2]: **\* [drivers/gpu/drm] Error 2
make[1]: **\* [drivers/gpu] Error 2
make: **\* [drivers] Error 2

Some referenced symbols and variables undefined in this version of nouveau source code.

**Comment by shinpei0208** [comment]
Okay I'll send you a set of my files that I'm using with 2.6.38.8.

**Comment by shinpei0208** [comment]
Also what would happen if you compile gdev/mod/drivers/nouveau?

**Comment by wkbjerry** [comment]
It fails to compile.

At first, it fails because it cannot find the header file <drm/drm_fourcc.h> file included in include/drm/drm_crtc.h. This is easy to solve by adding -I`pwd`/include to EXTRA_CFLAGS in the Makefile.

Then, it begins to complain not finding linux/export.h included in many c files under drivers/gpu/drm/. This is because linux 2.6.38.8 does not have a separate export.h file yet. All these errors can be solved by changing include linux/export.h to include linux/module.h.

After that, it begins to complain cannot find some functions referenced. For example, see this error message:

Makefile:622: "WARNING: Appending $KCPPFLAGS (-I/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/include/drm) from command line to kernel $CPPFLAGS"
/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/drivers/gpu/drm/nouveau/nouveau_bo.c: In function ‘nouveau_ttm_tt_populate’:
/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/drivers/gpu/drm/nouveau/nouveau_bo.c:1080: error: implicit declaration of function ‘swiotlb_nr_tbl’
make[3]: **\* [/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/drivers/gpu/drm/nouveau/nouveau_bo.o] Error 1
make[2]: **\* [/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/drivers/gpu/drm/nouveau] Error 2
make[1]: **\* [_module_/home/kaibo/projects/gdev/gdev/mod/nouveau-2.6.38.8/drivers/gpu/drm] Error 2
make: **\* [modules] Error 2

This is clear of a nouveau version not matching the 2.6.38.8 linux kernel.

**Comment by shinpei0208** [comment]
Crap... My 2.6.38.8 machine is down until next week. I will send you the module files as soon as possible when the machine is back online.

**Comment by wkbjerry** [comment]
That's fine. Thanks :)

**Comment by shinpei0208** [comment]
Damn it! It was my misunderstanding. The patch is not for 2.6.38.8, it is for 3.3.0...
For some reason, I was confused. Please install v3.3.0 and apply the patch.

**Comment by wkbjerry** [comment]
Got it! I'll try that and get back to you with the results :)

**Comment by wkbjerry** [comment]
My GTX480 arrived, and I tested 3.3.30 with gdev+nouveau (both patch and that bundled with gdev) as well as gdev+pscnv.
Both nouveau and pscnv now loads successfully. But, as I mentioned to you earlier, Shinpei, a new problem came out. I'll close this issue and open a new issue for the new problem.
